include/boost/beast/http/impl/read.ipp

```diff
@@ -112,72 +126,61 @@ operator()(
     error_code ec,
     std::size_t bytes_transferred)
 {
-    switch(state_)
+    size_t consumed = 0;
+    boost::optional<typename
+        DynamicBuffer::mutable_buffers_type> mb;
+
+    BOOST_ASIO_CORO_REENTER(*this)
     {
-    case 0:
-        state_ = 1;
-        if(b_.size() == 0)
-            goto do_read;
-        goto do_parse;
-
-    case 1:
-        state_ = 2;
-    case 2:
-        if(ec == boost::asio::error::eof)
+        if (b_.size() > 0)
         {
-            BOOST_ASSERT(bytes_transferred == 0);
-            if(p_.got_some())
+            consumed = p_.put(b_.data(), ec);
+            total_bytes_consumed_ += consumed;
+            b_.consume(consumed);
+            if (ec == http::error::need_more)
             {
-                // caller sees EOF on next read
                 ec.assign(0, ec.category());
-                p_.put_eof(ec);
-                if(ec)
-                    goto upcall;
-                BOOST_ASSERT(p_.is_done());
-                goto upcall;
             }
-            ec = error::end_of_stream;
-            goto upcall;
+            BOOST_ASIO_CORO_YIELD
+                s_.get_io_service().post(
+                    bind_handler(std::move(*this), ec, 0));
         }
-        if(ec)
-            goto upcall;
-        b_.commit(bytes_transferred);
-
-    do_parse:
-    {
-        auto const used = p_.put(b_.data(), ec);
-        bytes_transferred_ += used;
-        b_.consume(used);
-        if(! ec || ec != http::error::need_more)
-            goto do_upcall;
-        ec.assign(0, ec.category());
-    }
-
-    do_read:
-        try
-        {
-            mb_.emplace(b_.prepare(
-                read_size_or_throw(b_, 65536)));
-        }
-        catch(std::length_error const&)
+        while (total_bytes_consumed_ == 0 && !ec)
         {
-            ec = error::buffer_overflow;
-            goto do_upcall;
-        }
-        return s_.async_read_some(*mb_, std::move(*this));
-
-    do_upcall:
-        if(state_ >= 2)
-            goto upcall;
-        state_ = 3;
-        return s_.get_io_service().post(
-            bind_handler(std::move(*this), ec, 0));
+            mb = prepare_or_error(b_, ec);
+            if (ec)
+            {
+                BOOST_ASIO_CORO_YIELD
+                    s_.get_io_service().post(
+                        bind_handler(std::move(*this), ec, 0));
+                break;
```

Comment:
Is the meaning of `break` defined for stackless coroutines?

Reply:
```BOOST_ASIO_CORO_YIELD break;``` terminates the coroutine.
```break;``` outside of a yield statement refers to the logical block you're in (if you're in a loop, it will break out of it).